Advanced Panels: Add support for the 'Custom Fields' meta box

noisysocks opened the pull request:
Closes #3228.

![custom-fields](https://user-images.githubusercontent.com/612155/47542114-a5001c80-d927-11e8-851a-1fce1806021e.gif)

Adds an option which allows the user to enable the existing 'postcustom' meta box included in WordPress.

I've implemented this using the super hacky approach described in https://github.com/WordPress/gutenberg/pull/10676#issuecomment-430967687 😎 

The benefits of this approach are that:

- We can reuse the existing 'postcustom' meta box.
- The 'postcustom' assets aren't loaded unless the option is enabled.

The downsides are that:

- We have to store the option on the server (e.g. using a site option).
- The editor has to be reloaded when the option changes, which is pretty jarring.

noisysocks committed:
Advanced Panels: Add support for the 'Custom Fields' meta box

- Adds an option which allows the user to enable the existing
  'postcustom' meta box included in WordPress.
- Don't load the 'postcustom' assets when the option is disabled.
- Stores the option as a site option.
- Reload the editor when the option is changed.

noisysocks committed:
Fix E2E tests

The 'Enable Tips' checkbox is obscured by the first Tip now that the
Options modal is taller. Work around this by only testing that tips can
be re-enabled.

youknowriad reviewed (commented):
This makes sense unless we have REST API support for custom fields which I don't really expect it to come in 5.0 (or later?)

I'm adding some reviewers more familiar with the PHP side of things.

danielbachhuber commented on lib/meta-box-partial-page.php:
```diff
@@ -359,6 +365,32 @@ function the_gutenberg_metaboxes() {
+				if ( $('#postcustom').length ) {
+					$( '#the-list' ).wpList( {
+						addBefore: function( s ) {
+							s.data += '&post_id=' + $('#post_ID').val();
```
Should `$('#post_ID').val();` be cast to a number?

noisysocks replied:
I borrowed this from [Core](https://github.com/WordPress/wordpress-develop/blob/master/src/js/_enqueues/admin/post.js#L672). Looking at it some more, we could avoid the jQuery selector altogether by just interpolating `$post->ID`.

Fixed in 86b5b4e9a.

danielbachhuber commented on lib/meta-box-partial-page.php:
```diff
@@ -113,14 +113,20 @@ function gutenberg_filter_meta_boxes( $meta_boxes ) {
 
+	// Whether or not to load the 'postcustom' meta box is stored as a site option
+	// so that we're not always loading its assets.
+	$enable_custom_fields = get_option( 'enable_custom_fields', false );
```
Should this be `get_user_meta()` instead of `get_option()`? With the latter, it seems like you could have inconsistent behavior where User A could turn off the meta box for User B, and then User B turns it back on for User A, etc.

noisysocks replied:
You're totally right. Fixed in b02581c3e.

danielbachhuber commented on packages/edit-post/src/components/options-modal/options/enable-custom-fields.js:
```diff
@@ -0,0 +1,48 @@
+
+		// Redirect to an admin action that toggles the setting and reloads the editor
+		// with the meta box assets included on the page.
+		window.location.href = toggleURL;
```
Such a hack!

noisysocks replied:
Yeah, not proud of this! 🙃

noisysocks replied:
Thinking it might be nicer to run the action by calling `.submit()` on a hidden `<form>` that's on the page. That way, we'll trigger the action using a `POST` request instead of a `GET` request. (Not that we seem to mind in WP Admin, as e.g. the _Trash_ link deletes posts using `GET`.)

noisysocks replied:
I've done this in 6a18bcdb9f8e16135e5c15f595f18ab254f4d06b. I hate it a little less, now.

danielbachhuber commented on lib/client-assets.php:
```diff
@@ -1573,14 +1573,27 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 		'maxUploadFileSize'      => $max_upload_size,
```
It works!

![image](https://user-images.githubusercontent.com/36432/47562940-12cf2880-d8d5-11e8-8449-2d6bd7d730f4.png)

Should I be seeing that error message though?

youknowriad replied:
I saw the error mentioned for other meta boxes elsewhere. I wonder if it's a more generic bug in 5.0?

noisysocks replied:
How did you get this error?

danielbachhuber commented on lib/meta-box-partial-page.php:
```diff
@@ -402,3 +434,21 @@ function gutenberg_meta_box_post_form_hidden_fields( $post ) {
+	$enable_custom_fields = get_option( 'enable_custom_fields', false );
+	update_option( 'enable_custom_fields', ! $enable_custom_fields );
+
+	wp_safe_redirect( wp_get_referer() );
```
Needs an `exit` after `wp_safe_redirect()` to kill the process.

noisysocks replied:
👌fixed in fe4785d80.

danielbachhuber reviewed (dismissed):
Looking good! Few things to address.

noisysocks committed:
Custom Fields: Remove unnecessary jQuery selector

We can simply interpolate $post->ID when initializing Custom Fields,
rather than selecting from the #post_ID hidden field.

noisysocks committed:
Custom Fields: Use user meta field instead of site option

By using `get_user_meta()` instead of `get_option()`, the Custom Fields
setting won't affect other users.

noisysocks committed:
Custom Fields: Kill the process after redirecting

noisysocks committed:
Custom Fields: Use hidden <form> instead of redirect

Trigger the toggle_custom_fields action using a hidden <form> so that
the action is triggered by a POST request rather than a GET request.

noisysocks commented:
Noting to myself that since this touches the PHP we'll need to bring over these changes to the 5.0 branch.

noisysocks commented:
How do we feel about this? Would like to get it into the 5.0 RC.

talldan commented on packages/edit-post/src/components/options-modal/options/enable-custom-fields.js:
```diff
@@ -0,0 +1,47 @@
+
+		this.toggleCustomFields = this.toggleCustomFields.bind( this );
+
+		this.state = { isEnabled };
```
Nitpick: this state is called isChecked in DeferredOption.

noisysocks replied:
2c11c00ac6d9257b0df9ef63e7f8425906be9152

talldan reviewed (commented):
I had a look at the front-end side of things. The changes seem to be pretty minor there. Looks good to me - seems mergeable if the backend changes also look acceptable.

websupporter commented:
How does this approach take into account a custom post type where `custom-fields` are not supported, like for example

```
add_action('init', function() {
        remove_post_type_support( 'post', 'custom-fields' );
});
```

I would expect the option to not show up in the first place, but from a quick view into the code, I can't see a toggle for that.

In the classic editor, when I go to the screen options, the "Custom Fields"-option disappears when the support has been removed.

noisysocks committed:
Custom Fields: Move meta box options into their own component

noisysocks committed:
Custom Fields: Rename isEnabled → isChecked

noisysocks committed:
Custom Fields: Only show option if custom fields are supported

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -54,13 +59,21 @@ export function OptionsModal( { isModalActive, closeModal, metaBoxes = [] } ) {
-			) }
+			<Section title={ __( 'Advanced Panels' ) }>
+				<EnableCustomFieldsOption label={ __( 'Custom Fields' ) } />
+				{ map(
```
a7d0a0ca219cc42cc83388a3a8c97b2ae3a8dba0

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -54,13 +59,21 @@ export function OptionsModal( { isModalActive, closeModal, metaBoxes = [] } ) {
-				</Section>
-			) }
+			<Section title={ __( 'Advanced Panels' ) }>
+				<EnableCustomFieldsOption label={ __( 'Custom Fields' ) } />
```
Good catch. I've fixed this in 292485ad00208fe1526b0e0c6a5def27faed5c2d.

noisysocks commented:
Thanks all. I've responded to feedback and fixed the problem where _Custom Fields_ would display in post types that don't support `postcustom`.

One issue I've noticed: when the Advanced Custom Fields plugin is enabled, the _Custom Fields_ toggle is displayed even though it doesn't do anything. This is because the ACF plugin removes `postcustom` from the list of registered meta boxes. I don't think it's a blocker for this PR. Likely the fix is that ACF, when enabled, will have to mark the post type as not supporting `custom-fields`.

talldan reviewed (approved):
Yep, it's undergone a lot of reviews and feedback looks all addressed 👍

noisysocks commented:
🙈

noisysocks commented:
Lastly, created a ticket to remind ourselves to copy over the necessary PHP changes to Core: https://core.trac.wordpress.org/ticket/45257
